go/database/mpt/proof.go

```diff
@@ -0,0 +1,164 @@
+type rlpEncodedNode []byte
+
+// proofDb is a database of RLP encoded MPT nodes and their hashes that represent witness proofs.
+type proofDb map[common.Hash]rlpEncodedNode
+
+// WitnessProof represents a witness proof.
+// It contains a database of MPT nodes and their hashes.
+// The proof may be verified by iterating the proof database from the
+// root hash to a terminal entry according to a Path
+// representing either an account address or a storage key.
+// If the proof contains an account or a storage slot for the input address or key,
+// the proof is representing this account or storage.
+type WitnessProof proofDb
+
+// CreateWitnessProof creates a witness proof for the input account address
+// and possibly storage slots of the same account under the input storage keys.
+// If the proof cannot be created because either the address or a key does not exist,
+// it returns false.
+// It may return an error when it occurs in the underlying database.
+func CreateWitnessProof(nodeSource NodeSource, root *NodeReference, address common.Address, keys ...common.Key) (WitnessProof, bool, error) {
+	return nil, false, nil
+}
+
+// Add merges the input witness proof into the current witness proof.
+func (p WitnessProof) Add(other WitnessProof) {
+	for k, v := range other {
+		p[k] = v
+	}
+}
+
+// ExtractAccount extracts an account and storage from the current witness proof.
+// It returns a copy that contains only a proof for the input address and storage keys.
+// If the input address or keys are not found in the proof, it returns a proof that
+// contains only partial proof consisting only of elements that could be reached
+// from the input address and keys in the current proof.
+// If the proof cannot be fully extracted, this method returns false.
+func (p WitnessProof) ExtractAccount(root common.Hash, address common.Address, keys ...common.Key) (WitnessProof, bool) {
+	return WitnessProof{}, false
+}
+
+// IsValid checks that the proof has all valid entries.
+// It means, RLPs are valid encodings and their hashes match the key in the map.
+func (p WitnessProof) IsValid() bool {
+	return false
+}
+
+// GetAccountInfo extracts an account info from the witness proof for the input root hash and the address.
+// If the witness proof contains an account for the input address, it returns its information.
+// If the proof does not contain an account, it returns false.
+// The method may return an error if the proof is invalid.
+func (p WitnessProof) GetAccountInfo(root common.Hash, address common.Address) (AccountInfo, bool, error) {
+	path := AddressToNibblePath(address, nil) // TODO hash path without the node source
+	node := p.getNext(root, path)
+	if node == nil {
+		return AccountInfo{}, false, nil
+	}
+	return node.(*AccountNode).info, true, nil
+}
+
+// GetBalance extracts a balance from the witness proof for the input root hash and the address.
+// If the witness proof contains an account for the input address, it returns its balance.
```

Comment:
A proof may contain multiple accounts (since it may be the result of several "Add" calls). Thus, the documentation should consider this case:
```suggestion
// If the witness proof contains the requested account for the input address for the given root hash, it returns its balance.
```